node/src/components/chain_synchronizer/operations.rs

```diff
@@ -1289,11 +1295,199 @@ async fn fetch_block_worker(
+    for _ in 0..retries {
+        let maybe_signatures = ctx
+            .effect_builder
+            .fetch::<BlockSignatures>(block_hash, peer)
+            .await;
+        match maybe_signatures {
+            Ok(result) => return Ok(result),
+            Err(FetcherError::TimedOut { .. }) => continue,
+            Err(_) => return maybe_signatures,
+        }
+    }
+
+    Err(FetcherError::TimedOut {
+        id: block_hash,
+        peer,
+    })
+}
+
+async fn fetch_finality_signatures_by_block_header(
+    block_header: BlockHeader,
+    ctx: &ChainSyncContext<'_>,
+) -> Result<Option<BlockSignatures>, Error> {
+    let start = Timestamp::now();
+    let mut peer_list = ctx
+        .effect_builder
+        .get_fully_connected_non_joiner_peers()
+        .await;
+    ctx.filter_bad_peers(&mut peer_list);
+
+    let mut sig_collector = BlockSignaturesCollector::new();
+
+    let block_header_hash = block_header.hash(ctx.config.verifiable_chunked_hash_activation());
+
+    for peer in peer_list {
+        let fetched_signatures = fetch_finality_signatures_with_retry(
+            block_header_hash,
+            peer,
+            FINALITY_SIGNATURE_FETCH_RETRY_COUNT,
+            ctx,
+        )
+        .await;
+        let maybe_signatures = match fetched_signatures {
+            Ok(FetchedData::FromStorage { item, .. }) => {
+                trace!(
+                    ?block_header_hash,
+                    ?peer,
+                    "did not get FinalitySignatures from peer, got from storage instead",
+                );
+                ctx.metrics
+                    .observe_fetch_finality_signatures_duration_seconds(start);
+                return Ok(Some(*item));
+            }
+            Ok(FetchedData::FromPeer { item, .. }) => {
+                trace!(?block_header_hash, ?peer, "fetched FinalitySignatures");
+                Some(*item)
+            }
+            Err(_) => None,
+        };
+
+        let era_for_validators_retrieval = get_era_id_for_validators_retrieval(
+            &block_header.era_id(),
+            ctx.config.last_emergency_restart(),
+        );
+
+        let maybe_switch_block_of_previous_era = ctx
+            .effect_builder
+            .get_switch_block_header_at_era_id_from_storage(era_for_validators_retrieval)
+            .await;
+
+        let validator_weights = maybe_switch_block_of_previous_era
+            .as_ref()
+            .and_then(BlockHeader::next_era_validator_weights)
+            .ok_or(Error::HitGenesisBlockTryingToGetTrustedEraValidators {
+                trusted_header: block_header.clone(),
+            })?;
+        if let Some(signatures) = maybe_signatures {
+            consensus::validate_finality_signatures(&signatures, validator_weights)?;
+            sig_collector.add(signatures);
+
+            if sig_collector
+                .check_if_sufficient(validator_weights, ctx.config.finality_threshold_fraction())
```

Comment:
This is potentially wasteful if we end up calling `consensus::check_sufficient_finality_signatures` multiple times for new signatures. Hopefully not a bottleneck.

Reply:
Probably not a bottleneck, since `consensus::check_sufficient_finality_signatures` consists of simple arithmetic operations. Surely, there are loops, so this assumption is backed by the fact that we don't expect to have massive numbers of proofs or validators.